Support more np.random functions in ParallelAccelerator

ehsantn opened the pull request:
This PR adds support for analysis and parallelization of many np.random functions in ParallelAccelerator.

stuartarchibald commented on numba/tests/test_parfors.py:
```diff
@@ -434,6 +436,23 @@ def test_impl(v1, v2, m1, m2):
+            C = np.random.normal(0.0, 1.0, (n, n))
+            D = np.random.chisquare(1.0, (n, n))
+            E = np.random.randint(1, high=3, size=(n, n))
+            F = np.random.triangular(1, 2, 3, (n, n))
```
Perhaps this ought to include calls to all the distributions being added?

ehsantn replied:
For readability, only one call from each class is used. The control flow is identical for calls of each class (the recent refactor shows this better). We could get the lists and generate code here though.

stuartarchibald commented on numba/tests/test_parfors.py:
```diff
@@ -434,6 +436,23 @@ def test_impl(v1, v2, m1, m2):
+
+        n = 128
+        # only run parallel version and don't compare results since random
+        # numbers generated in parallel are different
```
Whilst not ideal, could an upper bound be calculated to assert some degree of validity on the answer?

ehsantn replied:
I added a check with 5% tolerance (ideally we need to use the statistical properties to find a tighter bound).

ehsantn commented on numba/ir_utils.py:
```diff
@@ -902,6 +902,54 @@ def simplify_CFG(blocks):
+                    saved_arr_arg[lhs] = arr
+                    scope = arr.scope
+                    loc = arr.loc
+                    # g_np_var = Global(numpy)
```
These kind of comments show what is being generate in the next couple of lines, which helps readability. Do you suggestions on other ways to do this?

stuartarchibald replied:
Ah, apologies, I was reading `Global` as a function and this as dead code, this is fine :)

ehsantn replied:
It's OK. Thanks.

ehsantn commented on numba/parfor.py:
```diff
@@ -456,7 +462,17 @@ def _is_supported_npycall(self, expr):
             return False
         call_name = self.array_analysis.numpy_calls[expr.func.name]
-        if call_name in ['zeros', 'ones', 'random.ranf']:
+        if call_name in ['zeros', 'ones', 'random.ranf', 'random.random_sample',
```
Thanks for the suggestion. I refactored the code and it looks much better now.

ehsantn commented:
Thank you for the comments @stuartarchibald . I applied them and also updated the docs. `full` and `empty` are not in the docs since they are not parallelized. Yes, `A.sum()` is supported now, so I updated the docs.

stuartarchibald commented:
Thanks for the fix-ups, much appreciated. Patch looks good. Merging.

ehsantn commented:
Thank you.
